### PR TITLE
[SPARK-15196][SparkR] Add a wrapper for dapply(repartition(col,...), ... )

### PR DESCRIPTION
**Pull request by NarineK**
## What changes were proposed in this pull request?

As mentioned in :
https://github.com/apache/spark/pull/12836#issuecomment-217338855
We would like to create a wrapper for: dapply(repartiition(col,...), ... )
This will allow to run aggregate functions on groups which are identified by a list of grouping columns.

I called the wrapper method gapply. We can rename it if we want to call it differently.
We could also have:
`setMethod("dapply", signature(x = "SparkDataFrame", func = "function", schema = "structType", col = "Column"),` , however, dapply already has many examples in the doc and if we add new examples for aggregate functions that will make the documentation longer and less clear.
## How was this patch tested?

Unit tests
1. Group by a column and compute mean
2.  Group by a column and train linear model

### CODE REVIEW

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -1214,6 +1214,77 @@ setMethod("dapply",
...
+#'---------
+#' Model  (Intercept)  Sepal_Width  Petal_Length  Petal_Width
+#' 1        0.699883    0.3303370    0.9455356    -0.1697527
+#' 2        1.895540    0.3868576    0.9083370    -0.6792238
+#' 3        2.351890    0.6548350    0.2375602     0.2521257
+#'
+#'}
+setMethod("gapply",
+          signature(x = "SparkDataFrame", func = "function", schema = "structType",
+                    col = "Column"),
```
Comment (shivaram) [inline R/pkg/R/DataFrame.R]: can we handle multiple columns now that we are using repartition ?

Comment (NarineK) [inline R/pkg/R/DataFrame.R]: yes, absolutely!

Comment (NarineK) [inline R/pkg/R/DataFrame.R]: I might have an issue in the signature, I'll fix it

**Comment by shivaram** [comment]
Change LGTM. Thanks @NarineK 

cc @sun-rui @felixcheung @davies - any other comments on this ?

**Comment by NarineK** [comment]
Thanks, @shivaram !
I have one question regarding factor datatypes. It seems that SparkR doesn't support it and we have to set: `stringsAsFactors = FALSE` in order to avoid "UnsupportedType" exceptions in data.frame.

Is it possible to map factor to maybe to list or string ... ?

R's data.frame converts strings to factor ... , but a factor in general doesn't have to be a string.

here is a snapshot from R's documentation: 

`Character variables passed to data.frame are converted to factor columns unless protected by I or argument stringsAsFactors is false. 
`
https://stat.ethz.ch/R-manual/R-devel/library/base/html/data.frame.html

**Comment by sun-rui** [comment]
@NarineK , @shivaram . sorry for missing the previous discussion. The problem is that repartition() can have multiple groups in a partition. This is not gapply() meant to handle with. Do I miss something?

**Comment by NarineK** [comment]
Hi @sun-rui , 
I think it depends on how we do the repartitioning. It shouldn't be the case when we do it by : 
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L2191
@davies , can it be the case ?

**Comment by sun-rui** [comment]
@NarineK, it is guaranteed that all items in a same group will be in a same partition. But  it is not guaranteed that there is only single group in a partition. There could be multiple groups in a partition.

**Comment by NarineK** [comment]
Thanks, @sun-rui !

Yes, It seems to be the case @sun-rui .
I've recently hit the case where the number of partitions was less than the number of actual groups.

I've tried the same thing on my previous implementation with groupByKey -> flatMap and it works perfectly fine with any repartitioning.

Maybe @davies, has some suggestions about this.  
If there is any repartitioner which will guarantee that each group will be in a single partition then we can use it otherwise this won't give us the expected result.

**Comment by davies** [comment]
@NarineK There is no such partitioner right now (it can't be cheap).

**Comment by NarineK** [comment]
Thanks, @davies !
This means that we cannot implement group-apply using repartitioning. 
What would you suggest in this case ?
My previous pull request works fine for one key.

We can try another implementation with groupBy-> agg . In this case I understand that we will need to implement imperative or declarative aggregate which most probably will collect the rows in a buffer with the same grouping column and pass that buffer with the rows to R worker.
Is this what you'd prefer ?

**Comment by davies** [comment]
@NarineK After repartition(), all the row with same grouping key are in the same partition, so we could have another groupBy() in R for single partition, then do the aggregate.

**Comment by NarineK** [comment]
I see, that's sounds good too, but, I'm not sure how user friendly it will be.
I guess for R side we need something like this:
df <- createDataFrame (sqlContext, iris)
schema <-  structType(structField("Species", "string"), structField("avg", "double"))
df <- repartition(df, col=df$"Species")
df1 <- dapply(
    df,
    function(x) {
        data.frame(aggregate(x$Sepal_Width, by = list(x$Species), FUN = mean), stringsAsFactors = FALSE)
    },
    schema)
collect(df1)

**Comment by NarineK** [comment]
What do you think, @shivaram , @sun-rui , @felixcheung  ?

**Comment by shivaram** [comment]
I am not sure why it should affect SparkR users. Users will still give us the same function to `gapply` as before but we will implement this inside SparkR side using the `aggregate` code snippet above ?

**Comment by NarineK** [comment]
Well, since the user provides the R function, I think he/she should provide the aggregate too. 
Instead of providing:
```
function(x) {
     data.frame(x$Species[1], mean(x$Sepal_Width), stringsAsFactors = FALSE)
 }
```

They will provide:
```
function(x) {
      data.frame(aggregate(x$Sepal_Width, by = list(x$Species), FUN = mean), stringsAsFactors = FALSE)
    }
```

This is my understanding, there may be also other ways ...

**Comment by NarineK** [comment]
This is similar to the following test case for repatition:
https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala#L1238

**Comment by sun-rui** [comment]
i suggest we follow the original design. i will take a detailed look at previous PR soon

**Comment by sun-rui** [comment]
@NarineK, I think we can close this PR?

**Comment by NarineK** [comment]
sure!

**Pull request closed without merging**